use right rounding method for floating point rounding

```
/home/marcus/src/gnuradio/gnuradio-runtime/lib/block_executor.cc:146:38: warning: casting (double + 0.5) to integer leads to incorrect rounding; consider using lround (#include <cmath>) instead [bugprone-incorrect-roundings]
                    new_tag.offset = ((double)new_tag.offset * rrate) + 0.5;

```

for large offsets, +0.5 can silently round to 0 ('doh).

might be more than related to #2055

more:
```
/home/marcus/src/gnuradio/gr-fec/lib/encoder_impl.cc:65:17: warning: casting (double + 0.5) to integer leads to incorrect rounding; consider using lround (#include <cmath>) instead [bugprone-incorrect-roundings]
    return (int)(0.5 + ninput * relative_rate());
                ^
/home/marcus/src/gnuradio/gr-fec/lib/encoder_impl.cc:70:17: warning: casting (double + 0.5) to integer leads to incorrect rounding; consider using lround (#include <cmath>) instead [bugprone-incorrect-roundings]
    return (int)(0.5 + noutput / relative_rate());
                ^
Suppressed 75 warnings (75 in non-user code).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
clang-tidy -header-filter=\.(c|cpp|h|hpp|cc)$ -export-fixes /tmp/tmpz5cv6ymt/tmpa4sjgruv.yaml -p=build /home/marcus/src/gnuradio/gr-fec/lib/repetition_encoder_impl.cc
81 warnings generated.
/home/marcus/src/gnuradio/gr-fec/lib/decoder_impl.cc:66:17: warning: casting (double + 0.5) to integer leads to incorrect rounding; consider using lround (#include <cmath>) instead [bugprone-incorrect-roundings]
    return (int)(0.5 + ninput * relative_rate());
                ^
/home/marcus/src/gnuradio/gr-fec/lib/decoder_impl.cc:71:17: warning: casting (double + 0.5) to integer leads to incorrect rounding; consider using lround (#include <cmath>) instead [bugprone-incorrect-roundings]
    return (int)(0.5 + noutput / relative_rate());
                ^
/home/marcus/src/gnuradio/gr-fec/lib/decoder_impl.cc:76:32: warning: casting (double + 0.5) to integer leads to incorrect rounding; consider using lround (#include <cmath>) instead [bugprone-incorrect-roundings]
    ninput_items_required[0] = 0.5 + fixed_rate_noutput_to_ninput(noutput_items);
                               ^
/home/marcus/src/gnuradio/gr-fec/lib/decoder_impl.cc:87:23: warning: casting (double + 0.5) to integer leads to incorrect rounding; consider using lround (#include <cmath>) instead [bugprone-incorrect-roundings]
    int outnum = (int)(((1.0 / relative_rate()) * noutput_items) + 0.5);
                      ^
/home/marcus/src/gnuradio/gr-fec/lib/decoder_impl.cc:89:14: warning: casting (double + 0.5) to integer leads to incorrect rounding; consider using lround (#include <cmath>) instead [bugprone-incorrect-roundings]
        (int)(relative_rate() * (ninput_items[0] - d_decoder->get_history()) + 0.5) /
             ^
/home/marcus/src/gnuradio/gr-fec/lib/decoder_impl.cc:110:9: warning: casting (double + 0.5) to integer leads to incorrect rounding; consider using lround (#include <cmath>) instead [bugprone-incorrect-roundings]
        items / relative_rate() * (output_multiple() - d_decoder->get_history()) + 0.5);

```

--- a/gnuradio-runtime/lib/block_executor.cc
+++ b/gnuradio-runtime/lib/block_executor.cc
@@ -143,7 +143,7 @@ static bool propagate_tags(block::tag_propagation_policy_t policy,
             } else if (use_fp_rrate) {
                 for (t = rtags.begin(); t != rtags.end(); t++) {
                     tag_t new_tag = *t;
-                    new_tag.offset = ((double)new_tag.offset * rrate) + 0.5;
+                    new_tag.offset = std::llround((double)new_tag.offset * rrate);
                     for (int o = 0; o < d->noutputs(); o++)
                         out_buf[o]->add_item_tag(new_tag);
                 }
@@ -191,7 +191,7 @@ static bool propagate_tags(block::tag_propagation_policy_t policy,
                 } else if (use_fp_rrate) {
                     for (t = rtags.begin(); t != rtags.end(); t++) {
                         tag_t new_tag = *t;
-                        new_tag.offset = ((double)new_tag.offset * rrate) + 0.5;
+                        new_tag.offset = std::llround((double)new_tag.offset * rrate);
                         out_buf->add_item_tag(new_tag);
                     }
                 } else {

--- a/gr-fec/lib/decoder_impl.cc
+++ b/gr-fec/lib/decoder_impl.cc
@@ -63,17 +63,17 @@ decoder_impl::decoder_impl(generic_decoder::sptr my_decoder,
 
 int decoder_impl::fixed_rate_ninput_to_noutput(int ninput)
 {
-    return (int)(0.5 + ninput * relative_rate());
+    return std::lround(ninput * relative_rate());
 }
 
 int decoder_impl::fixed_rate_noutput_to_ninput(int noutput)
 {
-    return (int)(0.5 + noutput / relative_rate());
+    return std::lround(noutput / relative_rate());
 }
 
 void decoder_impl::forecast(int noutput_items, gr_vector_int& ninput_items_required)
 {
-    ninput_items_required[0] = 0.5 + fixed_rate_noutput_to_ninput(noutput_items);
+    ninput_items_required[0] = std::lround(fixed_rate_noutput_to_ninput(noutput_items));
 }
 
 int decoder_impl::general_work(int noutput_items,
@@ -84,9 +84,9 @@ int decoder_impl::general_work(int noutput_items,
     const unsigned char* in = (unsigned char*)input_items[0];
     unsigned char* out = (unsigned char*)output_items[0];
 
-    int outnum = (int)(((1.0 / relative_rate()) * noutput_items) + 0.5);
+    int outnum = std::lround((1.0 / relative_rate()) * noutput_items);
     int innum =
-        (int)(relative_rate() * (ninput_items[0] - d_decoder->get_history()) + 0.5) /
+        std::lround(relative_rate() * (ninput_items[0] - d_decoder->get_history())) /
         (output_multiple() - d_decoder->get_history());
 
     int items = (outnum <= ninput_items[0] - d_decoder->get_history())
@@ -106,8 +106,8 @@ int decoder_impl::general_work(int noutput_items,
                      pmt::intern(alias()));
     }
 
-    int consumed = static_cast<int>(
-        items / relative_rate() * (output_multiple() - d_decoder->get_history()) + 0.5);
+    int consumed = std::lround(
+        items / relative_rate() * (output_multiple() - d_decoder->get_history()));
     int returned = items * (output_multiple() - d_decoder->get_history());
 
     consume_each(consumed);

--- a/gr-fec/lib/depuncture_bb_impl.cc
+++ b/gr-fec/lib/depuncture_bb_impl.cc
@@ -81,18 +81,18 @@ depuncture_bb_impl::~depuncture_bb_impl() {}
 
 int depuncture_bb_impl::fixed_rate_ninput_to_noutput(int ninput)
 {
-    return (int)(((d_puncsize / (double)(d_puncsize - d_puncholes)) * ninput) + .5);
+    return std::lround((d_puncsize / (double)(d_puncsize - d_puncholes)) * ninput);
 }
 
 int depuncture_bb_impl::fixed_rate_noutput_to_ninput(int noutput)
 {
-    return (int)((((d_puncsize - d_puncholes) / (double)(d_puncsize)) * noutput) + .5);
+    return std::lround(((d_puncsize - d_puncholes) / (double)(d_puncsize)) * noutput);
 }
 
 void depuncture_bb_impl::forecast(int noutput_items, gr_vector_int& ninput_items_required)
 {
     ninput_items_required[0] =
-        (int)((((d_puncsize - d_puncholes) / (double)(d_puncsize)) * noutput_items) + .5);
+        std::lround(((d_puncsize - d_puncholes) / (double)(d_puncsize)) * noutput_items);
 }
 
 /*
@@ -138,7 +138,7 @@ int depuncture_bb_impl::general_work(int noutput_items,
                  % ((int)(((1.0/relative_rate()) * noutput_items) + .5)));
     */
 
-    consume_each((int)(((1.0 / relative_rate()) * noutput_items) + .5));
+    consume_each(std::lround((1.0 / relative_rate()) * noutput_items));
     return noutput_items;
 }
 

--- a/gr-fec/lib/encoder_impl.cc
+++ b/gr-fec/lib/encoder_impl.cc
@@ -62,12 +62,12 @@ encoder_impl::~encoder_impl() {}
 
 int encoder_impl::fixed_rate_ninput_to_noutput(int ninput)
 {
-    return (int)(0.5 + ninput * relative_rate());
+    return std::lround(ninput * relative_rate());
 }
 
 int encoder_impl::fixed_rate_noutput_to_ninput(int noutput)
 {
-    return (int)(0.5 + noutput / relative_rate());
+    return std::lround(noutput / relative_rate());
 }
 
 void encoder_impl::forecast(int noutput_items, gr_vector_int& ninput_items_required)

--- a/gr-fec/lib/puncture_bb_impl.cc
+++ b/gr-fec/lib/puncture_bb_impl.cc
@@ -79,18 +79,18 @@ puncture_bb_impl::~puncture_bb_impl() {}
 
 int puncture_bb_impl::fixed_rate_ninput_to_noutput(int ninput)
 {
-    return (int)((((d_puncsize - d_puncholes) / (double)(d_puncsize)) * ninput) + .5);
+    return std::lround(((d_puncsize - d_puncholes) / (double)(d_puncsize)) * ninput);
 }
 
 int puncture_bb_impl::fixed_rate_noutput_to_ninput(int noutput)
 {
-    return (int)(((d_puncsize / (double)(d_puncsize - d_puncholes)) * noutput) + .5);
+    return std::lround((d_puncsize / (double)(d_puncsize - d_puncholes)) * noutput);
 }
 
 void puncture_bb_impl::forecast(int noutput_items, gr_vector_int& ninput_items_required)
 {
     ninput_items_required[0] =
-        (int)(((d_puncsize / (double)(d_puncsize - d_puncholes)) * noutput_items) + .5);
+        std::lround((d_puncsize / (double)(d_puncsize - d_puncholes)) * noutput_items);
 }
 
 /*
@@ -138,7 +138,7 @@ int puncture_bb_impl::general_work(int noutput_items,
                  % ((int)(((1.0/relative_rate()) * noutput_items) + .5)));
     */
 
-    consume_each((int)(((1.0 / relative_rate()) * noutput_items) + .5));
+    consume_each(std::lround((1.0 / relative_rate()) * noutput_items));
     return noutput_items;
 }
 

--- a/gr-fec/lib/puncture_ff_impl.cc
+++ b/gr-fec/lib/puncture_ff_impl.cc
@@ -79,18 +79,18 @@ puncture_ff_impl::~puncture_ff_impl() {}
 
 int puncture_ff_impl::fixed_rate_ninput_to_noutput(int ninput)
 {
-    return (int)((((d_puncsize - d_puncholes) / (double)(d_puncsize)) * ninput) + .5);
+    return std::lround(((d_puncsize - d_puncholes) / (double)(d_puncsize)) * ninput);
 }
 
 int puncture_ff_impl::fixed_rate_noutput_to_ninput(int noutput)
 {
-    return (int)(((d_puncsize / (double)(d_puncsize - d_puncholes)) * noutput) + .5);
+    return std::lround((d_puncsize / (double)(d_puncsize - d_puncholes)) * noutput);
 }
 
 void puncture_ff_impl::forecast(int noutput_items, gr_vector_int& ninput_items_required)
 {
     ninput_items_required[0] =
-        (int)(((d_puncsize / (double)(d_puncsize - d_puncholes)) * noutput_items) + .5);
+        std::lround((d_puncsize / (double)(d_puncsize - d_puncholes)) * noutput_items);
 }
 
 /*
@@ -138,7 +138,7 @@ int puncture_ff_impl::general_work(int noutput_items,
                  % ((int)(((1.0/relative_rate()) * noutput_items) + .5)));
     */
 
-    consume_each((int)(((1.0 / relative_rate()) * noutput_items) + .5));
+    consume_each(std::lround((1.0 / relative_rate()) * noutput_items));
     return noutput_items;
 }
 

